handler.go

```diff
@@ -45,40 +46,123 @@ func InputFromFile(filename string) (*Input, error) {
-					u.User = url.UserPassword(fm.Config.HubProxyUser, fm.Config.HubProxyPassword)
+				if len(fm.Config.HubProxyUser) > 0 {
+					proxyURL.User = url.UserPassword(fm.Config.HubProxyUser, fm.Config.HubProxyPassword)
 				}
-				tr.Proxy = func(_ *http.Request) (*url.URL, error) {
-					return u, nil
+				transport.Proxy = func(_ *http.Request) (*url.URL, error) {
+					return proxyURL, nil
 				}
 			}
 		}
+		fm.hubClient = &http.Client{
+			Timeout:   30 * time.Second,
+			Transport: transport,
+		}
+	})
+}
+
+// CheckHubCredentials performs credentials check for a Hub config, returning errors that reference
+// field names as in source config. Since config may be filled from file or UI, the field names can be different.
+// Consider also localization of UI, we want to decouple credential checking logic from their actual view in UI.
+//
+// Examples:
+// * for TOML: CheckHubCredentials(ctx, "hub_url", "hub_user", "hub_password")
+// * for WinUI: CheckHubCredentials(ctx, "URL", "User", "Password")
+func (fm *Frontman) CheckHubCredentials(ctx context.Context, fieldHubURL, fieldHubUser, fieldHubPassword string) error {
+	fm.initHubClientOnce()
+
+	if len(fm.Config.HubURL) == 0 {
+		return newEmptyFieldError(fieldHubURL)
+	} else if u, err := url.Parse(fm.Config.HubURL); err != nil {
+		err = errors.WithStack(err)
+		return newFieldError(fieldHubURL, err)
+	} else if u.Scheme != "http" && u.Scheme != "https" {
+		err := errors.Errorf("wrong scheme '%s', URL must start with http:// or https://", u.Scheme)
+		return newFieldError(fieldHubURL, err)
+	}
+	req, _ := http.NewRequest("HEAD", fm.Config.HubURL, nil)
+	req.Header.Add("User-Agent", fm.userAgent())
+	if len(fm.Config.HubUser) > 0 {
+		req.SetBasicAuth(fm.Config.HubUser, fm.Config.HubPassword)
+	}
+
+	ctx, cancelFn := context.WithTimeout(ctx, time.Minute)
+	req = req.WithContext(ctx)
+	resp, err := fm.hubClient.Do(req)
+	cancelFn()
+	if err = fm.checkClientError(resp, err, fieldHubUser, fieldHubPassword); err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
 
-		fm.hubHTTPClient = &http.Client{
-			Timeout:   time.Second * 30,
-			Transport: &tr,
+func (fm *Frontman) checkClientError(resp *http.Response, err error, fieldHubUser, fieldHubPassword string) error {
+	if err != nil {
+		if errors.Cause(err) == context.DeadlineExceeded {
+			err = errors.New("connection timeout, please check your proxy or firewall settings")
+			return err
+		}
+		return err
+	}
+	_, _ = io.Copy(ioutil.Discard, resp.Body)
+	resp.Body.Close()
+	if resp.StatusCode == http.StatusUnauthorized {
+		if len(fm.Config.HubUser) == 0 {
+			return newEmptyFieldError(fieldHubUser)
+		} else if len(fm.Config.HubPassword) == 0 {
+			return newEmptyFieldError(fieldHubPassword)
 		}
+		err := errors.Errorf("unable to authorize with provided Hub credentials (HTTP %d)", resp.StatusCode)
+		return err
+	} else if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusBadRequest {
```

Comment:
Can we omit the `else` since there is a `return` in every case inside the previous `if`?

Reply:
I have this code part shared and tested with Cagent, don't like to re-evaluate all ifs in my head twice.

Reply:
> don't like to re-evaluate all ifs in my head twice

uhm... ok

Reply:
This conditional block looks more compact with else if.

Reply:
> This conditional block looks more compact with else if.

I don't see how an additional `else`, which is not needed, makes it "look more compact"
but ok